app/services/device_notifier.rb

```diff
@@ -0,0 +1,77 @@
+class DeviceNotifier
+
+  # Used for testing notifications without talking to Apple services.
+  # Intended to be compatible with Houston::Client
+  class FakeAPNSClient
+    attr_reader :delivered_notifications
+
+    def initialize
+      @delivered_notifications = []
+    end
+
+    def push(payload)
+      @delivered_notifications << payload
+      Rails.logger.debug("Pushed fake notification_payload: #{payload.inspect}")
+    end
+  end
+
+  def self.notification_environment=(val)
+    new_environment = ActiveSupport::StringInquirer.new(val.to_s)
+    if new_environment.production? || new_environment.development? || new_environment.fake?
+      @@notification_environment = new_environment
+    else
+      raise ArgumentError.new("unknown notification environment: #{new_environment}")
+    end
+  end
+
+  def self.notification_environment
+    @@notification_environment
+  end
+
+  # It's important to regularly remove devices that Apple reports as stale. If
+  # you continue to try to notify stale devices, Apple may revoke your access
+  # to APNS!
+  def deactivate_stale_devices!
```

Comment:
You would run it with a cron job or something. The list of "stale devices" is obtained form Apple. They keep a log of the notifications that you have attempted to send to non-existant/stale devices, and have an endpoint where you can retrieve them.

The strategy that you were alluding to - wherein the users phone notifies you that their device token has changed - can be problematic if (e.g.) If Joe has a spot, and uninstalls the app (or just doesn't launch it for a long time, during which time his device token changes), then when Jane later captures that spot, I'm going to try to send notifications to Joe that he will never get. Apple will keep track of this, and later I can ask apple for the stale device tokens (`client.devices` is a (poorly named method) which calls to Apple)